pubsub.go

```diff
@@ -206,19 +245,21 @@ func (p *PubsubValueStore) rebroadcast(ctx context.Context) {
 		select {
 		case <-ticker.C:
 			var keys []string
+			var topics []*topicInfo
 			p.mx.Lock()
-			keys = make([]string, 0, len(p.subs))
-			for p, _ := range p.subs {
-				keys = append(keys, p)
+			topics = make([]*topicInfo, 0, len(p.topics))
+			for k, ti := range p.topics {
+				keys = append(keys, k)
+				topics = append(topics, ti)
 			}
 			p.mx.Unlock()
-			if len(keys) > 0 {
-				for _, k := range keys {
+			if len(topics) > 0 {
+				for i, k := range keys {
 					val, err := p.getLocal(k)
 					if err == nil {
-						topic := KeyToTopic(k)
+						topic := topics[i].topic
 						select {
-						case <-p.psPublishChannel(topic, val):
+						case <-p.psPublishChannel(ctx, topic, val):
```

Comment:
In theory we could emit an old record here and we should use a mutex or something to protect it. However, emitting old records isn't so bad and the network won't degrade much assuming infrequent rebroadcasts (which should be how this is used).

I'm also hoping we can largely just remove rebroadcasting entirely when PubSub starts emitting events for dropped messages.

I'm trying to minimize the number of changes that need review to get this approved, but if you'd like me to fix this lmk and I'll get right on it.

Reply:
Ah, you mean, given a concurrent publish, we could:

1. Start a rebroadcast tick.
2. Get an old value with `getLocal`.
3. Publish a new value (independently).
4. Re-publish the old value?

Yeah, that sounds fine. We could also just filter out strictly _worse_ values in the validator.

(FYI, providing event sequences like that makes understanding issues like this much easier. "This could emit old an old record" doesn't tell me how.)